src/log.js

```diff
@@ -240,80 +266,25 @@ class Log extends GSet {
-        return keys.find(keyMatches)
-      }
-      const pubkeys = this._keys.map(getPubKey)
+    // Get the difference of the logs
+    const newItems = Log.difference(log, this)
 
-      const verify = async (entry) => {
-        if (!entry.key) throw new Error("Entry doesn't have a public key")
-        if (!entry.sig) throw new Error("Entry doesn't have a signature")
-
-        if (this._keys.length === 1 && this._keys[0] === this._key ) {
-          if (entry.id !== this.id) 
-            throw new Error("Entry doesn't belong in this log (wrong ID)")
-        }
-
-        if (this._keys.length > 0 
-            && !this._keys.includes('*') 
-            && !checkAllKeys(this._keys.concat([this._key]), entry)) {
-          console.warn("Warning: Input log contains entries that are not allowed in this log. Logs weren't joined.")
-          return false
-        }
-
-        try {
-          await Entry.verifyEntry(entry, this._keystore)
-        } catch (e) {
-          throw new Error(`Invalid signature in entry '${entry.hash}'`)
-        }
-
-        return true
-      }
-
-      const checked = await pMap(entries, verify)
-      return checked.every(isTrue)
+    // Verify if entries are allowed to be added to the log and throws if
+    // there's an invalid entry
+    const permitted = async (entry) => {
+      const canAppend = await this._acl.canAppend(entry.key, entry)
+      if (!canAppend) throw new Error("Append not permitted")
     }
 
-    const difference = (log, exclude) => {
-      let stack = Object.keys(log._headsIndex)
-      let traversed = {}
-      let res = {}
-
-      const pushToStack = hash => {
-        if (!traversed[hash] && !exclude.get(hash)) {
-          stack.push(hash)
-          traversed[hash] = true
-        }
-      }
-
-      while (stack.length > 0) {
-        const hash = stack.shift()
-        const entry = log.get(hash)
-          if (entry && !exclude.get(hash) && entry.id === this.id) {
-          res[entry.hash] = entry
-          traversed[entry.hash] = true
-          entry.next.forEach(pushToStack)
-        }
-      }
-      return res
+    // Verify signature for each entry and throws if there's an invalid signature
+    const verify = async (entry) => {
+      const isValid = await Entry.verify(this._identity, entry)
+      if (!isValid) throw new Error(`Could not validate signature "${entry.sig}" for entry "${entry.hash}" and key "${entry.key}"`)
     }
 
-    // Merge the entries
-    const newItems = difference(log, this)
-
-    // if a key was given, verify the entries from the incoming log
-    if (this._key && this._key.getPublic) {
-      const canJoin = await verifyEntries(Object.values(newItems))
-      // Return early if any of the given entries didn't verify
-      if (!canJoin)
-        return this
-    }
+    const entriesToJoin = Object.values(newItems)
+    await pMap(entriesToJoin, permitted, { concurrency: 1 })
+    await pMap(entriesToJoin, verify, { concurrency: 1 })
```

Comment:
Really loving how much code we can remove here and simpler this whole function becomes! 👍❤️